### PR TITLE
feat(primitives): Remove Attributes

### LINKED ISSUE #510: fix(primitives): Remove Attributes Module
### Description

The [`attributes`](https://github.com/anton-rs/kona/blob/main/crates/primitives/src/attributes.rs) module in `kona-primitives` contains two core types to upstream into `op-alloy-rpc-types-engine`.
- `L2PayloadAttributes`
- `L2AttributesWithParent`

`op-alloy-rpc-types-engine` already defines a [`OptimismPayloadAttributes`](https://github.com/alloy-rs/op-alloy/blob/main/crates/rpc-types-engine/src/payload.rs#L10) that should serve as a drop-in replacement for `L2PayloadAttributes`, but `L2AttributesWithParent` will need to be upstreamed to `op-alloy-rpc-types-engine`.

Additionally, the `op-alloy-rpc-types-engine` should be made optionally `serde`, enabling the feature by default.

### PR DESCRIPTION
**Pull request by refcell**
### Description

Removes the `attributes` module from `kona-primitives`, re-using payload attributes types defined in `alloy-rpc-types-engine` and `op-alloy-rpc-types-engine`.

Closes #510

### CODE REVIEW

**Comment by refcell** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/anton-rs/kona/529?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#529** <a href="https://app.graphite.dev/github/pr/anton-rs/kona/529?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#528** <a href="https://app.graphite.dev/github/pr/anton-rs/kona/528?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @refcell and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: Cargo.lock [hunk 1]
```diff
@@ -2910,6 +2930,18 @@ dependencies = [
...
+name = "op-alloy-rpc-types-engine"
+version = "0.2.11"
+source = "git+https://github.com/alloy-rs/op-alloy?branch=main#95fdc87239d58079c8db2226779545f10758ba77"
+dependencies = [
+ "alloy-primitives",
+ "alloy-rpc-types-engine",
+ "alloy-serde",
+ "op-alloy-protocol",
+ "serde",
+]
```
Comment (refcell) [inline Cargo.lock:2943]: Unfortunately, this is why the build on cannon is failing. `alloy-serde` is not currently hidden behind the `serde` feature flag so it leaks serde and std features into the `no_std` environment.

File: Cargo.lock [hunk 2]
```diff
@@ -504,6 +504,21 @@ dependencies = [
...
+checksum = "1c031a91e94a39f928244bc837c953817be5b8cc61759e1a9123b3abd17560dd"
+dependencies = [
+ "alloy-consensus",
+ "alloy-eips",
+ "alloy-primitives",
+ "alloy-rlp",
+ "alloy-serde",
+ "derive_more 1.0.0",
+ "serde",
+]
```
Comment (refcell) [inline Cargo.lock:520]: The issue is in upstream `alloy-rpc-types-engine` as well. Working on fixing this in pr https://github.com/alloy-rs/alloy/pull/1284

File: crates/derive/src/stages/attributes_queue.rs [hunk 3]
```diff
@@ -143,8 +146,15 @@ where
             .prepare_payload_attributes(parent, batch.epoch())
             .await
             .map_err(StageError::AttributesBuild)?;
-        attributes.no_tx_pool = true;
-        attributes.transactions.extend(batch.transactions);
+        attributes.no_tx_pool = Some(true);
```
Comment (clabby) [inline crates/derive/src/stages/attributes_queue.rs:149]: Why is this an option in `OptimismPayloadAttributes`? Should always be there, doesn't have to be backwards compat w/ L1.

File: crates/derive/src/stages/attributes_queue.rs [hunk 4]
```diff
@@ -143,8 +146,15 @@ where
...
+        match attributes.transactions {
+            Some(ref mut txs) => txs.extend(batch.transactions),
+            None => {
+                if !batch.transactions.is_empty() {
+                    attributes.transactions = Some(batch.transactions);
+                }
+            }
+        }
```
Comment (clabby) [inline crates/derive/src/stages/attributes_queue.rs:157]: same same - why is this an option?

**Review by clabby — CHANGES_REQUESTED** [review]
Looks good, but I'd prefer if we changed `OptimismPayloadAttributes` to not have `Option`s for `transactions`, `no_tx_pool`. These should always be there, since the type is not intended to be backwards-compat w/ L1 payload attributes.

**Comment by refcell** [comment]
> Looks good, but I'd prefer if we changed `OptimismPayloadAttributes` to not have `Option`s for `transactions`, `no_tx_pool`. These should always be there, since the type is not intended to be backwards-compat w/ L1 payload attributes.

They're all marked `omitempty` in the reference implementation: https://github.com/ethereum-optimism/optimism/blob/8dd6fb36aad7a1cd5c9e7e877b340b2e0c5cc14c/op-service/eth/types.go#L326-L331

Do you want to diverge from this for any specific reason?

**Review by clabby — APPROVED** [review]

**Pull request merged**